Welcome to the final project of “Apache Spark for Scalable Machine Learning on BigData”. In this assignment you’ll analyze a real-world dataset and apply machine learning on it using Apache Spark. 

In order to pass, you need to implement some code (as described in the instruction section on Coursera) and finally answer a quiz on the Coursera platform.

Let’s start by downloading the dataset and creating a dataframe. This dataset can be found on DAX, the IBM Data Asset Exchange and can be downloaded for free.

https://developer.ibm.com/exchanges/data/all/jfk-weather-data/


In [1]:
from pyspark import SparkContext, SparkConf
from pyspark.sql import SQLContext, SparkSession
from pyspark.sql.types import StructType, StructField, DoubleType, IntegerType, StringType

sc = SparkContext.getOrCreate(SparkConf().setMaster("local[*]"))

from pyspark.sql import SparkSession

spark = SparkSession \
    .builder \
    .getOrCreate()

spark

In [2]:
# delete files from previous runs
!rm -f jfk_weather*

# download the file containing the data in CSV format
!wget http://max-training-data.s3-api.us-geo.objectstorage.softlayer.net/noaa-weather/jfk_weather.tar.gz

# extract the data
!tar xvfz jfk_weather.tar.gz

--2020-03-20 12:09:09--  http://max-training-data.s3-api.us-geo.objectstorage.softlayer.net/noaa-weather/jfk_weather.tar.gz
Resolving max-training-data.s3-api.us-geo.objectstorage.softlayer.net (max-training-data.s3-api.us-geo.objectstorage.softlayer.net)... 67.228.254.196
Connecting to max-training-data.s3-api.us-geo.objectstorage.softlayer.net (max-training-data.s3-api.us-geo.objectstorage.softlayer.net)|67.228.254.196|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2575759 (2.5M) [application/x-tar]
Saving to: ‘jfk_weather.tar.gz’

jfk_weather.tar.gz  100%[===================>]   2.46M  1.55MB/s    in 1.6s    

2020-03-20 12:09:12 (1.55 MB/s) - ‘jfk_weather.tar.gz’ saved [2575759/2575759]

./._jfk_weather.csv
jfk_weather.csv


In [3]:
# create a dataframe out of it by using the first row as field names and trying to infer a schema based
# on contents
df = spark.read.option("header", "true").option("inferSchema","true").csv('jfk_weather.csv')

# register a corresponding query table
df.createOrReplaceTempView('df')

The dataset contains some null values, therefore schema inference doesn’t work properly for all columns, in addition, a column contained trailing characters, so we need to clean up the data set first.  
This is a normal task in any data science project since your data is never clean, don’t worry if you don’t understand all code, you won’t be asked about it. 

In [4]:
import random
random.seed(42)

from pyspark.sql.functions import translate, col

df_cleaned = df \
    .withColumn("HOURLYWindSpeed", df.HOURLYWindSpeed.cast('double')) \
    .withColumn("HOURLYWindDirection", df.HOURLYWindDirection.cast('double')) \
    .withColumn("HOURLYStationPressure", translate(col("HOURLYStationPressure"), "s,", "")) \
    .withColumn("HOURLYPrecip", translate(col("HOURLYPrecip"), "s,", "")) \
    .withColumn("HOURLYRelativeHumidity", translate(col("HOURLYRelativeHumidity"), "*", "")) \
    .withColumn("HOURLYDRYBULBTEMPC", translate(col("HOURLYDRYBULBTEMPC"), "*", ""))

df_cleaned = df_cleaned \
                .withColumn("HOURLYStationPressure", df_cleaned.HOURLYStationPressure.cast('double')) \
                .withColumn("HOURLYPrecip", df_cleaned.HOURLYPrecip.cast('double')) \
                .withColumn("HOURLYRelativeHumidity", df_cleaned.HOURLYRelativeHumidity.cast('double')) \
                .withColumn("HOURLYDRYBULBTEMPC", df_cleaned.HOURLYDRYBULBTEMPC.cast('double'))

df_filtered = df_cleaned.filter("""
    HOURLYWindSpeed <> 0
    and HOURLYWindSpeed IS NOT NULL
    and HOURLYWindDirection <> 0
    and HOURLYWindDirection IS NOT NULL
    and HOURLYStationPressure <> 0
    and HOURLYStationPressure IS NOT NULL
    and HOURLYPressureTendency <> 0
    and HOURLYPressureTendency IS NOT NULL
    and HOURLYPressureTendency <> 0
    and HOURLYPressureTendency IS NOT NULL
    and HOURLYPrecip <> 0
    and HOURLYPrecip IS NOT NULL
    and HOURLYRelativeHumidity <> 0
    and HOURLYRelativeHumidity IS NOT NULL
    and HOURLYDRYBULBTEMPC <> 0
    and HOURLYDRYBULBTEMPC IS NOT NULL
""")

We want to predict the value of one column based of some others. It is sometimes helpful to print a correlation matrix. 

In [5]:
from pyspark.ml.feature import VectorAssembler

vectorAssembler = VectorAssembler(inputCols=["HOURLYWindSpeed", ],
                                  outputCol="features")
df_pipeline = vectorAssembler.transform(df_filtered)


#
from pyspark.ml.stat import Correlation

Correlation.corr(df_pipeline, "features").head()[0].toArray()

array([[ 1.        ,  0.06306013, -0.4204518 ],
       [ 0.06306013,  1.        , -0.19199348],
       [-0.4204518 , -0.19199348,  1.        ]])

In [18]:
# Course Project Quiz - Q1
vectorAssembler_Q1 = VectorAssembler(inputCols=["HOURLYWindSpeed", "HOURLYPressureTendency"],
                                  outputCol="features")
df_pipeline_Q1 = vectorAssembler_Q1.transform(df_filtered)


#
# from pyspark.ml.stat import Correlation
Correlation.corr(df_pipeline_Q1, "features").head()[0].toArray()


array([[1.        , 0.06432947],
       [0.06432947, 1.        ]])

As we can see:  
*HOURLYWindSpeed* and *HOURLYWindDirection* correlate with 0.06306013 whereas  
*HOURLYWindSpeed*  and *HOURLYStationPressure* correlate with -0.4204518,  
this is a good sign if we want to predict *HOURLYWindSpeed* from *HOURLYWindDirection* and *HOURLYStationPressure*.

Since this is supervised learning, let’s split our data into train (80%) and test (20%) set.

In [6]:
splits = df_filtered.randomSplit([0.8, 0.2])

df_train = splits[0]
df_test = splits[1]

Again, we can re-use our feature engineering pipeline

In [7]:
from pyspark.ml.feature import StringIndexer, OneHotEncoder
from pyspark.ml.linalg import Vectors
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.feature import Normalizer
from pyspark.ml import Pipeline

vectorAssembler = VectorAssembler(inputCols=[
  "HOURLYWindDirection",
  "ELEVATION",
  "HOURLYStationPressure"], outputCol="features")

normalizer = Normalizer(inputCol="features", outputCol="features_norm", p=1.0)

Now we define a function for evaluating our regression prediction performance.  
We’re using RMSE (Root Mean Squared Error) here, the smaller the better...

In [8]:
from pyspark.ml.evaluation import RegressionEvaluator

def regression_metrics(prediction):        
    evaluator = RegressionEvaluator(
    labelCol="HOURLYWindSpeed", predictionCol="prediction", metricName="rmse")
    rmse = evaluator.evaluate(prediction)
    print("RMSE on test data = %g" % rmse)
    return rmse

Let’s run a linear regression model first for building a baseline.



In [9]:
# LR1
from pyspark.ml.regression import LinearRegression

lr = LinearRegression(labelCol="HOURLYWindSpeed", featuresCol='features', 
                      maxIter=100, regParam=0.0, elasticNetParam=0.0)

pipeline = Pipeline(stages=[vectorAssembler, normalizer, lr])
model = pipeline.fit(df_train)
prediction = model.transform(df_test)

regression_metrics(prediction)

RMSE on test data = 6.03474


6.0347415133058835

In [19]:
# Course Project Quiz - Q3
# Please change #LR1 in order to use features_norm over features. What's the RMSE value you get now?
lr_Q3 = LinearRegression(labelCol="HOURLYWindSpeed", featuresCol='features_norm', 
                      maxIter=100, regParam=0.0, elasticNetParam=0.0)


pipeline_Q3 = Pipeline(stages=[vectorAssembler, normalizer, lr_Q3])
model_Q3 = pipeline_Q3.fit(df_train)
prediction_Q3 = model_Q3.transform(df_test)

regression_metrics(prediction_Q3)

RMSE on test data = 1.03543


1.03543219955495

Now we’ll try a Gradient Boosted Tree Regressor

In [10]:
# GBT1
from pyspark.ml.regression import GBTRegressor

gbt = GBTRegressor(labelCol="HOURLYWindSpeed", maxIter=100)

pipeline = Pipeline(stages=[vectorAssembler, normalizer, gbt])
model = pipeline.fit(df_train)
prediction = model.transform(df_test)

regression_metrics(prediction)

RMSE on test data = 6.588


6.58800459944986

Now let’s switch gears. Previously, we tried to predict HOURLYWindSpeed, but now we want predict HOURLYWindDirection.  
In order to turn this into a classification problem we discretize the value using the Bucketizer. The new feature is called HOURLYWindDirectionBucketized.

In [13]:
from pyspark.ml.feature import Bucketizer, OneHotEncoder

bucketizer = Bucketizer(splits=[0, 180, float('Inf') ], 
                        inputCol="HOURLYWindDirection", outputCol="HOURLYWindDirectionBucketized")

encoder = OneHotEncoder(inputCol="HOURLYWindDirectionBucketized", outputCol="HOURLYWindDirectionOHE")

Again, we define a function in order to assess how we perform. Here we just use the accuracy measure which gives us the fraction of correctly classified examples.  
Here, 0 is bad, 1 is good.

In [14]:
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

def classification_metrics(prediction):    
    mcEval = MulticlassClassificationEvaluator()\
      .setMetricName("accuracy")\
      .setPredictionCol("prediction")\
      .setLabelCol("HOURLYWindDirectionBucketized")
    accuracy = mcEval.evaluate(prediction)
    print("Accuracy on test data = %g" % accuracy)
    return accuracy

Again, for baselining we use LogisticRegression.

In [15]:
# LGReg1 - LogRegr
from pyspark.ml.classification import LogisticRegression

lr = LogisticRegression(labelCol="HOURLYWindDirectionBucketized", maxIter=10)
#,"ELEVATION","HOURLYStationPressure","HOURLYPressureTendency","HOURLYPrecip"

vectorAssembler = VectorAssembler(inputCols=["HOURLYWindSpeed","HOURLYDRYBULBTEMPC"],
                                  outputCol="features")

pipeline = Pipeline(stages=[bucketizer, vectorAssembler, normalizer, lr])
model = pipeline.fit(df_train)
prediction = model.transform(df_test)

classification_metrics(prediction)

Accuracy on test data = 0.595318


0.5953177257525084

Let’s try some other Algorithms and see if model performance increases.  
It’s also important to tweak other parameters like parameters of individual algorithms (e.g. number of trees for RandomForest) or parameters in the feature engineering pipeline, e.g. train/test split ratio, normalization, bucketing, ...

In [16]:
# RF1
from pyspark.ml.classification import RandomForestClassifier

rf = RandomForestClassifier(labelCol="HOURLYWindDirectionBucketized", numTrees=30)

vectorAssembler = VectorAssembler(inputCols=["HOURLYWindSpeed", "HOURLYDRYBULBTEMPC", 
                                             "ELEVATION", "HOURLYStationPressure",
                                             "HOURLYPressureTendency", "HOURLYPrecip"],
                                  outputCol="features")


pipeline = Pipeline(stages=[bucketizer,vectorAssembler,normalizer,rf])
model = pipeline.fit(df_train)
prediction = model.transform(df_test)

classification_metrics(prediction)

Accuracy on test data = 0.658863


0.6588628762541806

In [20]:
# Course Project Quiz - Q8
# If you change the number of trees in cell #RF1 from 30 to 10, what's the new accuracy?

rf_Q08 = RandomForestClassifier(labelCol="HOURLYWindDirectionBucketized", numTrees=10)

vectorAssembler_Q08 = VectorAssembler(inputCols=["HOURLYWindSpeed", "HOURLYDRYBULBTEMPC", 
                                             "ELEVATION", "HOURLYStationPressure",
                                             "HOURLYPressureTendency", "HOURLYPrecip"],
                                  outputCol="features")


pipeline_Q08 = Pipeline(stages=[bucketizer, vectorAssembler, normalizer, rf_Q08])
model_Q08 = pipeline_Q08.fit(df_train)
prediction_Q08 = model_Q08.transform(df_test)

classification_metrics(prediction_Q08)

Accuracy on test data = 0.648829


0.6488294314381271

In [17]:
# GBT2
from pyspark.ml.classification import GBTClassifier

gbt = GBTClassifier(labelCol="HOURLYWindDirectionBucketized", maxIter=100)

vectorAssembler = VectorAssembler(inputCols=["HOURLYWindSpeed", "HOURLYDRYBULBTEMPC",
                                             "ELEVATION", "HOURLYStationPressure",
                                             "HOURLYPressureTendency","HOURLYPrecip"],
                                  outputCol="features")

pipeline = Pipeline(stages=[bucketizer,vectorAssembler,normalizer,gbt])

model = pipeline.fit(df_train)
prediction = model.transform(df_test)

classification_metrics(prediction)

Accuracy on test data = 0.64214


0.6421404682274248